### PR TITLE
enh: Retrieve and store token usage in execution metadata

### PR DESCRIPTION
**Pull request by tdraier**
## Description

Add token_usage in chat and llm blocks completion result, containing  number of prompt/completion/total tokens. The data are sent in the meta block : 

```
{
    "type": "block_execution",
    "content": {
        "block_type": "chat",
        "block_name": "MODEL",
        "execution": [
            [
                {
                    "value": {
                        "message": {
                            "role": "assistant",
                            "content": "Hello!"
                        }
                    },
                    "error": null,
                    "meta": {
                        "logs": [],
                        "token_usage": {
                            "prompt_tokens": 9,
                            "completion_tokens": 5,
                            "total_tokens": 14
                        }
                    }
                }
            ]
        ]
    }
}
```

Also send a token_usage event when using streaming ( the previous block_execution is already sent , I'm not sure this event is very useful ) : 

```
{
    "type": "token_usage",
    "content": {
        "block_type": "chat",
        "block_name": "MODEL",
        "input_index": 0,
        "map": null,
        "token_usage": {
            "prompt_tokens": 9,
            "completion_tokens": 5,
            "total_tokens": 14
        }
    }
}
```

Currently implemented for openAI , mistral, anthropic

## Risk

No real risk, it's new data that were not previously sent. 
Completely safe to rollback

### CODE REVIEW

File: core/src/providers/openai.rs [hunk 1]
```diff
@@ -1149,11 +1171,12 @@ pub async fn streamed_chat_completion(
                     finish_reason: None,
                 })
                 .collect::<Vec<_>>(),
+            usage,
         };
 
         for i in 0..guard.len() {
             let a = guard[i].clone();
-            if a.choices.len() != f.choices.len() {
+            if a.choices.len() != f.choices.len() && a.usage.is_none() {
```
Comment (spolu) [inline core/src/providers/openai.rs:1179]: Why this condition?

**Review by spolu — COMMENTED** [review]
Fantastic work

Given the blast radius of the PR I'd like to test it tomorrow before LGTM

We can remove the event as well as discussed + left a question

File: core/src/providers/openai.rs (hunk 1 above)
Comment (tdraier) [reply core/src/providers/openai.rs:1179]: that one is not very clear .. when requesting the token usage, we receive a dedicated fake chunk with only usages, but no completion/choices .. and as f is a clone of the first chunk and contain one choice, this test fails if we use it on that usage chunk. another solution would be to not add the chunk if it does not contain any choice (line 1094)

Comment (tdraier) [reply core/src/providers/openai.rs]: updated code to test before adding chunk

**Review by spolu — APPROVED** [review]
LGTM

**Pull request merged**